### PR TITLE
Implement RFC 1058

### LINKED ISSUE #26906: Rename slice::{tail, init} per RFC 1058
Tracking issue for https://github.com/rust-lang/rfcs/pull/1058

### PR DESCRIPTION
**Pull request by nagisa**
Fixes #26906

### CODE REVIEW

**Comment by nagisa** [comment]
Still waiting for tests to finish and hunting for last occurences of `tail` and `init`.

File: src/libcollections/slice.rs [hunk 1]
```diff
@@ -280,34 +280,34 @@ impl<T> [T] {
         core_slice::SliceExt::first_mut(self)
     }
 
-    /// Returns all but the first element of a slice.
-    #[unstable(feature = "slice_extras", reason = "likely to be renamed")]
+    /// Returns the first and all the rest elements of a slice.
```
Comment (apasel422) [inline src/libcollections/slice.rs]: s/rest elements/rest of the elements/

File: src/libcollections/slice.rs [hunk 2]
```diff
@@ -280,34 +280,34 @@ impl<T> [T] {
...
-    pub fn tail(&self) -> &[T] {
-        core_slice::SliceExt::tail(self)
+    pub fn split_first(&self) -> Option<(&T, &[T])> {
+        core_slice::SliceExt::split_first(self)
     }
 
-    /// Returns all but the first element of a mutable slice
-    #[unstable(feature = "slice_extras",
-               reason = "likely to be renamed or removed")]
+    /// Returns the first and all the rest elements of a slice.
```
Comment (apasel422) [inline src/libcollections/slice.rs]: Same here.

File: src/libcollections/slice.rs [hunk 3]
```diff
@@ -280,34 +280,34 @@ impl<T> [T] {
...
-        core_slice::SliceExt::tail_mut(self)
+    pub fn split_first_mut(&mut self) -> Option<(&mut T, &mut [T])> {
+        core_slice::SliceExt::split_first_mut(self)
+
     }
 
-    /// Returns all but the last element of a slice.
-    #[unstable(feature = "slice_extras", reason = "likely to be renamed")]
+    /// Returns the last and all the rest elements of a slice.
+    #[stable(feature = "rfc1058", since = "1.3.0")]
```
Comment (apasel422) [inline src/libcollections/slice.rs]: Same here.

File: src/libcollections/slice.rs [hunk 4]
```diff
@@ -280,34 +280,34 @@ impl<T> [T] {
...
-        core_slice::SliceExt::init(self)
+    pub fn split_last(&self) -> Option<(&T, &[T])> {
+        core_slice::SliceExt::split_last(self)
+
     }
 
-    /// Returns all but the last element of a mutable slice
-    #[unstable(feature = "slice_extras",
-               reason = "likely to be renamed or removed")]
+    /// Returns the last and all the rest elements of a slice.
```
Comment (apasel422) [inline src/libcollections/slice.rs]: Same here.

File: src/libcollections/slice.rs (hunk 1 above)
Comment (nagisa) [inline src/libcollections/slice.rs]: Thanks!

File: src/libcollections/slice.rs [hunk 5]
```diff
@@ -280,34 +280,34 @@ impl<T> [T] {
         core_slice::SliceExt::first_mut(self)
     }
 
-    /// Returns all but the first element of a slice.
-    #[unstable(feature = "slice_extras", reason = "likely to be renamed")]
+    /// Returns the first and all the rest of the elements of a slice.
+    #[stable(feature = "rfc1058", since = "1.3.0")]
```
Comment (alexcrichton) [inline src/libcollections/slice.rs]: This is a fairly significant change in terms of functionality so I don't think we should insta-stabilize these methods (e.g. leave them as `#[unstable]` for a little longer). Could you also give it a more descriptive feature name than "rfc1058"? Just to help error messages and such later on down the road.

Finally, could you leave the old `tail` methods and such, but `#[deprecate]` them in favor of these new methods?

**Comment by alexcrichton** [comment]
Thanks @nagisa!

**Comment by nagisa** [comment]
Comments addressed.

**Comment by alexcrichton** [comment]
@bors: r+ 7a90865db5b9bbf4e94580a886e3ab187d91c83c

Thanks!

**Pull request merged**